packages/ui-extensions/src/surfaces/admin/api/order-routing-rule/metafields.ts

```diff
@@ -0,0 +1,54 @@
+const supportedDefinitionTypes = [
+  'boolean',
+  'collection_reference',
+  'color',
+  'date',
+  'date_time',
+  'dimension',
+  'file_reference',
+  'json',
+  'metaobject_reference',
+  'mixed_reference',
+  'money',
+  'multi_line_text_field',
+  'number_decimal',
+  'number_integer',
+  'page_reference',
+  'product_reference',
+  'rating',
+  'rich_text_field',
+  'single_line_text_field',
+  'product_taxonomy_value_reference',
+  'url',
+  'variant_reference',
+  'volume',
+  'weight',
+  'list.collection_reference',
+  'list.color',
+  'list.date',
+  'list.date_time',
+  'list.dimension',
+  'list.file_reference',
+  'list.metaobject_reference',
+  'list.mixed_reference',
+  'list.number_decimal',
+  'list.number_integer',
+  'list.page_reference',
+  'list.product_reference',
+  'list.rating',
+  'list.single_line_text_field',
+  'list.url',
+  'list.variant_reference',
+  'list.volume',
+  'list.weight',
+] as const;
+
+interface MetafieldChange {
```

Comment:
@olavoasantos, @MitchLillie, there's still no easy way to expose this to the extensions (nor to web, for that matter).

I managed to resolve the type through some TS gymnastics like:

```ts
type ExtractElementType<T> = T extends readonly (infer ElementType)[]
  ? ElementType
  : never;

type ApiForRenderExtension = ReturnType<
  typeof useApi<'admin.settings.order-routing-rule.render'>
>;

type ExtractMethod<T extends keyof ApiForRenderExtension> =
  ApiForRenderExtension[T];

type SaveMetafields = ExtractMethod<'saveMetafields'>;

type MetafieldChange = ExtractElementType<Parameters<SaveMetafields>[0]>;
```

I like how the types are clean and properly derived from `useApi(TARGET)`, but wondering if we'd want to expose some of the helpers that already exist in here as well in order to make this slightly simpler...

Normally one wouldn't need this, but as things grow, types might be needed in files that do not call `useApi` making some of this necessary 🤔

Reply:
> I like how the types are clean and properly derived from useApi(TARGET)

Couldn't you just export and import these types directly? I know this would make them not-perfectly derived from useApi, but it would make the extension code simpler.

Reply:
Yep, makes sense. This would be a good enhancement. I created https://github.com/Shopify/ui-extensions/issues/1525